### DISCUSSION TITLE
Make (System.Linq) IPartition<TElement> and IIListProvider<TElement> public

### DISCUSSION THREAD

**Discussion by chamalulu** [dotnet/runtime#70390] [Ideas] [answered]
Could these two interfaces be made public for third party code to implement s.t. LINQ can use these optimizations?
Are they not stable enough?
Are there other reasons they are internal?

**Comment by huoyaoyuan** [comment, marked as answer]
They are meant to represent intermediate results of LINQ, as a part of implementation detail.
We don't want to expose such implementation detail.

> Are they not stable enough?

They are just introduced quite recently, far before the original LINQ implementation.

**Reply by chamalulu** [reply]
I see. Thank you for the answer.
I recognize you need to tread carefully w.r.t. public API surface.
Hypothetically, though, it would be useful to be able to implement, for example, IPartition<TElement> on a type when you can provide a constant time implementation of TryGetLast where Enumerable.Last otherwise would fall back on a linear time operation for your type.
